internals/daemon/api_notices.go

```diff
@@ -0,0 +1,178 @@
+		noticeType := state.NoticeType(typeStr)
+		if !noticeType.Valid() {
+			return statusBadRequest("invalid notice type %q", typeStr)
+		}
+		types = append(types, noticeType)
+	}
+
+	keys := strutil.MultiCommaSeparatedList(query["keys"])
+
+	afterStr := query.Get("after")
+	var after time.Time
+	if afterStr != "" {
+		var err error
+		after, err = time.Parse(time.RFC3339, afterStr)
+		if err != nil {
+			return statusBadRequest("invalid after timestamp %q: %v", afterStr, err)
+		}
+	}
+
+	filter := &state.NoticeFilter{
+		Types: types,
+		Keys:  keys,
+		After: after,
+	}
+	var notices []*state.Notice
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	timeoutStr := query.Get("timeout")
+	if timeoutStr != "" {
+		// Wait up to timeout for notices matching given filter to occur
+		timeout, err := time.ParseDuration(timeoutStr)
+		if err != nil {
+			return statusBadRequest("invalid timeout %q: %v", timeoutStr, err)
+		}
+
+		ctx, cancel := context.WithTimeout(r.Context(), timeout)
+		defer cancel()
+
+		notices, err = st.WaitNotices(ctx, filter)
+		if errors.Is(err, context.Canceled) {
+			return statusBadRequest("request canceled")
+		}
+		// DeadlineExceeded will occur if timeout elapses; in that case return
+		// an empty list of notices, not an error.
+		if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+			return statusInternalError("cannot wait for notices: %s", err)
+		}
+	} else {
+		// No timeout given, fetch currently-available notices
+		notices = st.Notices(filter)
+	}
+
+	if len(notices) == 0 {
+		notices = []*state.Notice{} // avoid null result
+	}
+	return SyncResponse(notices)
+}
+
+func v1PostNotices(c *Command, r *http.Request, _ *UserState) Response {
+	var payload struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after"`
+		Data        map[string]string `json:"data"`
+	}
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return statusBadRequest("cannot decode request body: %v", err)
+	}
+
+	if payload.Action != "add" {
+		return statusBadRequest("invalid action %q", payload.Action)
+	}
+	if payload.Type != "custom" {
+		return statusBadRequest(`invalid type %q (can only add "custom" notices)`, payload.Type)
+	}
+	if !customKeyRegexp.MatchString(payload.Key) {
+		return statusBadRequest(`invalid key %q (must be in "domain.com/key" format)`, payload.Key)
+	}
+	if len(payload.Key) > maxNoticeKeyLength {
+		return statusBadRequest("key must be %d bytes or less", maxNoticeKeyLength)
+	}
+
+	var repeatAfter time.Duration
+	if payload.RepeatAfter != "" {
+		var err error
+		repeatAfter, err = time.ParseDuration(payload.RepeatAfter)
+		if err != nil {
+			return statusBadRequest("invalid repeat-after duration %q: %v", payload.RepeatAfter, err)
+		}
+	}
+
+	dataSize := 0
+	for k, v := range payload.Data {
```

Comment:
If we unmarshal Data separately into a json.RawMessage, we can look into it before decoding.

Reply:
I don't really understand why that's better: I get that it means we could compare the data size directly (though that would include JSON punctuation and whitespace too), but then we have to unmarshal again after that to get the map.

Reply:
Per voice discussion, we'll use a RawMessage per suggestion above and compare the size of that. It's only going to get smaller if anything (whitespace removed etc).